### ISSUE TITLE
Support deferred messages or delayed processing

### ISSUE THREAD

**Issue by fgheysels** [arcus-azure/arcus.messaging#411]
We might have processes where we receive a message from servicebus that cannot be processed yet.  (For instance because we're waiting on another message coming from another party first).
In such situations, it would be good if we could defer the message or resubmit the same message.
It would be nice if we could do this in the implementation of the message-handler.  I can see that the `AzureServiceBusMessageHandler` now already has functionality to Complete or Abandon messages via protected methdos.  Some 'defer' or 'resubmit' logic there would be appropriate I think.

I think there are 2 approaches:

## Re-submitting:
I'm thinking of having the possibility of re-submitting the exact same message to Service Bus again, but with a specific / specified 
 `ScheduledEnqueueTimeUtc` property. [Docs on this here](https://learn.microsoft.com/en-us/dotnet/api/microsoft.azure.servicebus.message.scheduledenqueuetimeutc?view=azure-dotnet).  (That would probably mean completing the initial message and then resubmitting a copy of it , with all the same properties etc... but with a specified ScheduledEnqueueTimeUtc ?
(Or can this be implemented by abandoning the message and passing in the necessary 'propertiesToModify' ?  [ pass in `ScheduledEnqueueTimeUtc` as propertyToModify ? )

## Deferring

This might be another possible approach, but I think it's more complex?
Some documentation on the defer-functionality that is provided by Azure ServiceBus can be found [here](https://learn.microsoft.com/en-us/azure/service-bus-messaging/message-deferral).
The drawback of deferring -as I see it-, is that you need another mechanism to retrieve deferred messages again, and that you need to specify which deferred message you want to retrieve again (so you need to keep state / administration for this).

**Comment by stijnmoreels** [comment]
That would be a cool feature, indeed. I see some similarity with #25 and #36 .

**Comment by fgheysels** [comment]
> That would be a cool feature, indeed. I see some similarity with #25 and #36 .

I see some similarity with #25 indeed.  However, @tomkerkhove mentions there that the pump should orchestrate this, but I don't know how the pump could do that.  The pump has no knowledge of the 'business logic' that must be executed (what must be executed first).
Therefore, my assumption would be that the message-handler would be responsible for 'resubmitting' the message if it cannot be processed yet.

**Comment by fgheysels** [comment]
I've tried implementing this by calling the `AbandonAsync` method and specify a `scheduledEnqueueTimeUtc` property while abandoning, but this doesn't give us the desired results.

This issue is related to what we want to achieve here: https://github.com/Azure/azure-service-bus/issues/454

One of the workarounds that is proposed in the above issue is to 'Defer' the message and keep track of the deferred message by sending another message to the queue which contains the sequence-number of the deferred message so that it can be retrieved later.
I think this might be a good approach for a work around.

I was thinking of having something like:

- a (protected) method in `AzureServiceBusMessageHandler`  that you can call to 'Defer' or re-submit a message using a method that has a signature like this:  `DeferMessageAsync( DateTimeOffset  resubmitDateTime);`.
  When this method is called it should:
  - Defer the incoming message using the `ProcessMessageEventArgs.DeferMessageAsync` method.
  - Post a new message on ServiceBus with the necesary meta-data (amongst others the SequenceNumber of the received message that was deferred.   The `scheduledEnqueueTimeUtc` property of this new message must be set to the `resubmitDateTime` value that was passed as a parameter.

- When the pump / messagerouter receives the message that has been posted because of the deferral, the messagerouter should be able to retrieve the original deffered message via the `ReceiveDeferredMessageAsync` [method](https://learn.microsoft.com/en-us/dotnet/api/azure.messaging.servicebus.servicebusreceiver.receivedeferredmessageasync?view=azure-dotnet#azure-messaging-servicebus-servicebusreceiver-receivedeferredmessageasync(system-int64-system-threading-cancellationtoken)) and the `sequenceNumber` that was stored in that message.  Once the deferred message is retrieved, it should be routed to the correct MessageHandler, just as we're doing now with 'normal' messages.

**Comment by stijnmoreels** [comment]
Sounds like a plan! 🎉 Thx a lot for documenting/investigating this so thoroughly. Will see if I can work on something while you're gone. 👍

**Comment by stijnmoreels** [comment]
The feature is done being designed and almost picked up for development: https://github.com/Azure/azure-service-bus/issues/454#issuecomment-1645816895
Maybe we can wait a bit for it.

**Comment by fgheysels** [comment]
Well, I think this is something we're going to need very soon in a project.  Also, other teams might find it very useful.
I don't know how soon MSFT will start development, and how soon it will be released / available.   I guess we can close that gap for now using our 'work around implementation', and then, once MSFT has provided a cleaner solution, modify our implementation to make use of the feature that MSFT has developed ?

**Comment by stijnmoreels** [comment]
Yes, think that is a good approach. Unless of course this is very critical for other projects, otherwise I would be in favor to wait a bit for Microsoft to catch up.

**Comment by fgheysels** [comment]
I think it is a crucial feature, something that should have been implemented by MSFT from the start imho.
I'm afraid it can take up to 6 months or more before MSFT will deploy this.

Another thing regarding the possible implementation.  I've discussed this with @gverstraete yesterday as well, and his first idea on how to implement this, could also be an option:

When 'deferring' a message, we could create a clone of the message, (with the same body and properties) and send this message to service bus with a `ScheduledEnqueuetimeUtc', and next to that we must then also make sure to 'accept' the original message so that it is no longer available in the queue.

**Comment by stijnmoreels** [comment]
Ah yes, but I'm wondering what Impact this could have on the performance/scaling/cost of Service Bus as for every message, there will be two in the queue.

**Comment by fgheysels** [comment]
> Ah yes, but I'm wondering what Impact this could have on the performance/scaling/cost of Service Bus as for every message, there will be two in the queue.

Service Bus bills you per operation, not by the amount of messages.  [Service Bus Pricing](https://azure.microsoft.com/en-us/pricing/details/service-bus/).
Of course, in a situation where you need to retrieve a deferred message, that 's an additional operation, but I would say that additional costs for this would be marginal ?

> Otherwise, this is indeed a rather elegant solution. If we're thinking about implementing this, we should also probably note somewhere that the authorization of the Service Bus resource will require Write (new messages) too, instead of Read only.

That's a good point, and indeed a drawback.

**Comment by fgheysels** [comment]
The thing is that we'll need to decide on how to continue with this:
- implement using deferring the message and enqueue a message that says 'retrieve the deferred message'
- implement via the approach that 'clones' the original message and sends it with a scheduled date, abandon the original message
- wait for MSFT

I would be in favor to not to wait for MSFT as we don't know how soon the feature will be delivered.   We can close the gap for now with an alternative approach.   Once MSFT comes up with their solution, we can rework our workaround to use the 'official implementation'.

What's your take @gverstraete ? 
This is a feature we'll certainly require at a project I'm working on with @jcools85 right now, so he might have some ideas as well ?

**Comment by gverstraete** [comment]
> The thing is that we'll need to decide on how to continue with this:
> 
> * implement using deferring the message and enqueue a message that says 'retrieve the deferred message'
> * implement via the approach that 'clones' the original message and sends it with a scheduled date, abandon the original message
> * wait for MSFT
> 
> I would be in favor to not to wait for MSFT as we don't know how soon the feature will be delivered. We can close the gap for now with an alternative approach. Once MSFT comes up with their solution, we can rework our workaround to use the 'official implementation'.
> 
> What's your take @gverstraete ? This is a feature we'll certainly require at a project I'm working on with @jcools85 right now, so he might have some ideas as well ?

Agree, I think cloning makes most sense

**Comment by jcools85** [comment]
Any update on this? We need this feature at NxtPort to implement a retry mechanism with custom backoff times.

**Comment by stijnmoreels** [comment]
> Any update on this? We need this feature at NxtPort to implement a retry mechanism with custom backoff times.

We are working on a circuit breaker, but that would not include deferred messaging.
See also [this feature request](https://github.com/Azure/azure-service-bus/issues/454) on Microsoft's Service Bus, as there is some missing functionlity.

No time/budget to implement a custom solution as of yet.

**Comment by stijnmoreels** [comment]
I think it is possible, though, to defer the message yourself in the message handler, add an application property with a timestamp, and filter with another message handler for certain past times. But as I've said, no room to take anything else at the moment.

**Comment by fgheysels** [comment]
I think it is very hard to implement this in the MessageHandler itself.

It looks like MSFT is not very responsible on my requests in [this issue](https://github.com/Azure/azure-service-bus/issues/454) so my plan for a workaround in Arcus would be like this:

1. Introduce a `DeferMessageAsync(TimeSpan delay)` method.  This method would call the `DeferMessageAsync` method of the `ServiceBusReceiver
2. Since we're now responsible for retrieving the deferred message ourselves; the `DeferMessageAsync` method of Arcus should send a scheduled message to servicebus which contains the details of the deferred message.  (It definitely needs the `SequenceNumber` of the deferred message).
3. An Arcus ServiceBus messagepump must have a default `MessageHandler` (or another mechanism) for receiving and processing the scheduled message that we've sent in step 2.  This mechanism must make sure that it can interpret the 
message, retrieve the stored `SequenceNumber` and retrieve the deferred message via the `SequenceNumber` using the `ReceiveDeferredMessage` method of the `ServiceBusReceiver`.
4. The Arcus MessagePump would then also need to make sure that the received deferred message is sent to the Arcus `MessageRouter` so that the correct registered message-handler can process the message.

More info on deferring in ServiceBus can be found here: https://learn.microsoft.com/en-us/azure/service-bus-messaging/message-deferral#message-deferral-apis.   The approach I mentionned here is actually more or less described in the 'sample scenarios' paragraph of that article.

**Comment by stijnmoreels** [comment]
We're talking about the same thing here 😉

**Comment by fgheysels** [comment]
We cannot do that in the messagehandler.
We currently have no access to the `ServiceBusMessage` itself, nor do we have a way to make sure that the deferred message that we would then retrieve, is send through the message-router.

**Comment by fgheysels** [comment]
Since the feature in the ServiceBus SDK [is being delayed](https://github.com/Azure/azure-service-bus/issues/454#issuecomment-2675213529), I think we might have to come up with an alternative ourselves here.

I think it can be a rather simple workaround, were we provide a method on the `AzureServiceBusMessageContext` which accepts a TimeSpan.
The implementation of this method calls the `DeferMessageAsync` method that exists on the `ServiceBusMessageReceiver`.  The `MessageId` of the deferred message must be packed into an 'envelope' (a message structure that we define ourselves?) and that 'envelope' must be posted to ServiceBus with a `ScheduledEnqueueTime` (using the TimeSpan that is passed as an argument).
The MessageRouter must have a MessageHandler registered that is able to accept those 'DeferredMessageEnvelope' messages.  Upon receival of such a message, the deferred message -identified by the MessageId that is mentioned inside the 'DeferredMessageEnvelope' must be retrieved via the `ReceiveDeferredMessageAsync` method and that message must be given to the MessageRouter.

**Comment by stijnmoreels** [comment]
Think with #551 , this workaround has become a lot easier to do. Something to looked at for v3.0.

**Comment by fgheysels** [comment]
MSFT should normally release an updated version of their SDK by the end of this year that contains this feature, so maybe we should just wait for it and then implement it using the functionality that is brought by the SDK ?
